### ISSUE TITLE
What has to be done to get this running in cf

### ISSUE THREAD

**Issue by kmacpher67** [pivotal-topher-bullock/x11-boshrelease#1]
What's the trickery for this? 
Does this make all the headache of this stuff go away? 
https://github.com/cloudfoundry/nodejs-buildpack/issues/89
Is this ready for public usage? 
We require a X.11 that's near impossible to get working on standard non-root apt-get system.

**Comment by topherbullock** [comment]
Definitely hoping to remove a lot of the headaches around getting X11 installed and running. Using `bosh-gen` following [this doc](http://www.starkandwayne.com/blog/bosh-packages-with-debian/ ) from @drnic on packaging apt packages with BOSH to get around the nasty looking build process for Xorg, and version blobs properly.

It uses `dpkg` to unpack the .deb packages produced by `apt-get`-ing the aptitude packages we want to the install target. Not sure offhand what the implication of doing that in the buildpack compile phase would be, or what the exact limitation of having "non-root apt-get" is for dpkg.

I've definitely been focused on the BOSH parts, and its not quite for public consumption yet. Its a flex-hour project for me so its in the early stages. Extending CF buildpacks or some other packaging step for getting it into a droplet for CF is on my radar.

**Comment by kmacpher67** [comment]
if you look at the heroku solution, it is basically `dpkg` using `apt-get` but overriding the target directories: 
Take look at the `bin/compile` you see the apt-get stuff in there: 
https://github.com/kmacpher67/nodejs-buildpack/tree/xvfb2/bin

`APT_OPTIONS="-o debug::nolocking=true -o dir::cache=$APT_CACHE_DIR -o dir::state=$APT_STATE_DIR"` 

`apt-get $APT_OPTIONS -y --force-yes -d install --reinstall $PACKAGE `
basically either by using the apt-get cache based on the o/s or pulling the deb file. 
Either way, this adds a considerable amount of time to the build startup process. 

The other key element is the sym linking and environment PATHS, LD_LIB_PATH, Etc. 

Is it worth while to have this put into a standard build-pack or pull/pushed into the cloudfoundry repos?

**Comment by topherbullock** [comment]
@kmacpher67 
Yeah, I just pushed and cut a BOSH release which puts all the debs into a cache and runs `apt-get install` with the cache in the job which runs `Xorg`. Since this is a BOSH release there will be some work involved in adding it to a CF buildback, but the idea behind the packaging should be largely the same. 

I don't think X11 belongs in the officical nodeJS buildpack, but what I would suggest is making a custom buildpack forked off the official nodeJS buildpack. If you keep the surface area of the changes light you should be able to apply your additional X11 stuff on top of any new changes to the official nodeJS buildpack, so that your custom one can stay up to date with patches / CVEs.

**Comment by topherbullock** [comment]
@kmacpher67 Closing this in favor of creating [a separate science project repo](https://github.com/pivotal-topher-bullock/x11-buildpack.git) focused on CF Buildpack addons for X11 😄 

Hopefully this is something that can just be scripted on top of _any_ official buildpack.

**Comment by kmacpher67** [comment]
```should be able to apply your additional X11 stuff on top of any new changes to the official nodeJS buildpack, so that your custom one can stay up to date with patches / CVEs.```
I was explaining this to somebody and I realize the way I forked the repo, that my changes are a dead end branch! 
I'm sorry I'm new at this, not good at git submodule trickery and the like. 
Do you know where a how-to for dummies where I can walk through making a more "inherited" xvfb instead of a dead end branch? 

- `x11-buildpack` 
The ideas here is we add it to the list in my inherited xvfb branch and the X.11 dependencies are there. When you say X.11 do you also envision the xvfb magic?
